File: backend-java/src/main/java/com/neria/manager/chat/ChatAuthService.java
package com.neria.manager.chat;

import com.neria.manager.common.entities.ChatUser;
import com.neria.manager.common.repos.ChatUserRepository;
import com.neria.manager.common.services.ScryptHasher;
import io.jsonwebtoken.Claims;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class ChatAuthService {
  private final ChatUserRepository chatUserRepository;
  private final ScryptHasher scryptHasher;
  private final ChatTokenService tokenService;
  private final String salt;

  public ChatAuthService(
      ChatUserRepository chatUserRepository, ScryptHasher scryptHasher, ChatTokenService tokenService) {
    this.chatUserRepository = chatUserRepository;
    this.scryptHasher = scryptHasher;
    this.tokenService = tokenService;
    String envSalt = System.getenv().getOrDefault("CHAT_PASSWORD_SALT", "");
    if (envSalt.length() < 16) {
      throw new IllegalArgumentException("CHAT_PASSWORD_SALT must be at least 16 characters");
    }
    this.salt = envSalt;
  }

  public String hashPassword(String value) {
    return scryptHasher.hash(value, salt);
  }

  private boolean matches(String value, String hash) {
    return scryptHasher.matches(value, salt, hash);
  }

  public Map<String, Object> register(String tenantId, RegisterChatUserRequest dto) {
    String email = dto.email == null ? "" : dto.email.trim().toLowerCase();
    if (email.isBlank()) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email required");
    }
    chatUserRepository.findByTenantIdAndEmail(tenantId, email).ifPresent(existing -> {
      throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Email already registered");
    });

    ChatUser user = new ChatUser();
    user.setId(UUID.randomUUID().toString());
    user.setTenantId(tenantId);
    user.setEmail(email);
    user.setName(dto.name != null ? dto.name.trim() : null);
    user.setPasswordHash(hashPassword(dto.password));
    user.setStatus("active");
    user.setCreatedAt(LocalDateTime.now());
    user.setUpdatedAt(LocalDateTime.now());
    chatUserRepository.save(user);
    return issueToken(user);
  }

  public Map<String, Object> login(String tenantId, LoginChatUserRequest dto) {
    String email = dto.email == null ? "" : dto.email.trim().toLowerCase();
    ChatUser user =
        chatUserRepository
            .findByTenantIdAndEmail(tenantId, email)
            .orElseThrow(
                () -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials"));
    if (!"active".equals(user.getStatus())) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
    }
    if (!matches(dto.password, user.getPasswordHash())) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
    }
    return issueToken(user);
  }

  public Map<String, Object> issueToken(ChatUser user) {
    Map<String, Object> claims = new HashMap<>();
    claims.put("sub", user.getId());
    claims.put("tenantId", user.getTenantId());
    claims.put("email", user.getEmail());
    String token = tokenService.sign(claims);
    Map<String, Object> userView =
        Map.of(
            "id", user.getId(),
            "tenantId", user.getTenantId(),
            "email", user.getEmail(),
            "name", user.getName(),
            "status", user.getStatus());
    return Map.of("accessToken", token, "user", userView);
  }

  public Claims validateToken(String token) {
    try {
      return tokenService.verify(token);
    } catch (Exception ex) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid chat token");
    }
  }

  public ChatUser getUserById(String id) {
    return chatUserRepository.findById(id).orElse(null);
  }

  public static class RegisterChatUserRequest {
    public String email;
    public String name;
    public String password;
  }

  public static class LoginChatUserRequest {
    public String email;
    public String password;
  }
}

File: backend-javascript/src/chat/chat-auth.service.ts
import { ForbiddenException, Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { InjectRepository } from '@nestjs/typeorm';
import { scryptSync, timingSafeEqual } from 'crypto';
import { Repository } from 'typeorm';
import { ChatUser } from '../common/entities/chat-user.entity';
import { RegisterChatUserDto } from './dto/register-chat-user.dto';
import { LoginChatUserDto } from './dto/login-chat-user.dto';

@Injectable()
export class ChatAuthService {
  private readonly salt: string;

  constructor(
    private readonly jwtService: JwtService,
    @InjectRepository(ChatUser)
    private readonly chatUsersRepository: Repository<ChatUser>
  ) {
    const salt = process.env.CHAT_PASSWORD_SALT || '';
    if (salt.length < 16) {
      throw new Error('CHAT_PASSWORD_SALT must be at least 16 characters');
    }
    this.salt = salt;
  }

  private hash(value: string) {
    return scryptSync(value, this.salt, 32).toString('hex');
  }

  private compare(value: string, hash: string) {
    const computed = this.hash(value);
    return timingSafeEqual(Buffer.from(computed, 'hex'), Buffer.from(hash, 'hex'));
  }

  hashPassword(value: string) {
    return this.hash(value);
  }

  async register(tenantId: string, dto: RegisterChatUserDto) {
    const existing = await this.chatUsersRepository.findOne({
      where: { tenantId, email: dto.email.toLowerCase() }
    });
    if (existing) {
      throw new ForbiddenException('Email already registered');
    }

    const user = this.chatUsersRepository.create({
      tenantId,
      email: dto.email.toLowerCase(),
      name: dto.name?.trim() || null,
      passwordHash: this.hash(dto.password)
    });
    await this.chatUsersRepository.save(user);

    return this.issueToken(user);
  }

  async login(tenantId: string, dto: LoginChatUserDto) {
    const user = await this.chatUsersRepository.findOne({
      where: { tenantId, email: dto.email.toLowerCase() }
    });
    if (!user || user.status !== 'active') {
      throw new UnauthorizedException('Invalid credentials');
    }
    if (!this.compare(dto.password, user.passwordHash)) {
      throw new UnauthorizedException('Invalid credentials');
    }
    return this.issueToken(user);
  }

  async issueToken(user: ChatUser) {
    const payload = {
      sub: user.id,
      tenantId: user.tenantId,
      email: user.email
    };
    const accessToken = await this.jwtService.signAsync(payload);
    return {
      accessToken,
      user: {
        id: user.id,
        tenantId: user.tenantId,
        email: user.email,
        name: user.name,
        status: user.status
      }
    };
  }

  async validateToken(token: string) {
    try {
      return await this.jwtService.verifyAsync(token);
    } catch (err) {
      throw new UnauthorizedException('Invalid chat token');
    }
  }

  async getUserById(id: string) {
    return this.chatUsersRepository.findOne({ where: { id } });
  }
}
